test/jdk/tools/jar/JarExtractTest.java

```diff
@@ -0,0 +1,382 @@
+        }
+        return abs;
+    }
+
+    @Test(dataProvider = "relExtractLocations")
+    public void testExtractToRelativeDir(final String dest) throws Exception {
+        testExtract(dest);
+    }
+
+    @Test(dataProvider = "absExtractLocations")
+    public void testExtractToAbsoluteDir(final String dest) throws Exception {
+        testExtract(dest);
+    }
+
+    @Test(dataProvider = "absNormalizedExtractLocations")
+    public void testExtractToAbsoluteNormalizedDir(final String dest) throws Exception {
+        testExtract(dest);
+    }
+
+    /**
+     * Test that {@code jar -x -f --dir} works as expected
+     */
+    @Test
+    public void testExtractLongForm() throws Exception {
+        final String dest = "foo-bar";
+        System.out.println("Extracting " + testJarPath + " to " + dest);
+        final int exitCode = JAR_TOOL.run(System.out, System.err, "-x", "-f", testJarPath.toString(),
+                "--dir", dest);
+        Assert.assertEquals(exitCode, 0, "Failed to extract " + testJarPath + " to " + dest);
+        verifyExtractedContent(dest);
+    }
+
+    /**
+     * Verifies that the {@code jar --help} output contains the --dir option
+     */
+    @Test
+    public void testHelpOutput() {
+        final ByteArrayOutputStream outStream = new ByteArrayOutputStream();
+        final int exitCode = JAR_TOOL.run(new PrintStream(outStream), System.err, "--help");
+        Assert.assertEquals(exitCode, 0, "jar --help command failed");
+        final String output = outStream.toString();
+        final String expectedMsg = "--dir                    Directory into which the jar will be extracted";
+        Assert.assertTrue(output.contains(expectedMsg), "jar --help didn't contain --dir option");
+    }
+
+    /**
+     * Tests that {@code jar -x -f} command works fine even when the -C or --dir option
+     * isn't specified
+     */
+    @Test
+    public void testExtractWithoutOutputDir() throws Exception {
+        final int exitCode = JAR_TOOL.run(System.out, System.err, "-x", "-f", testJarPath.toString());
+        Assert.assertEquals(exitCode, 0, "Failed to extract " + testJarPath);
+        // the content would have been extracted to current dir
+        verifyExtractedContent(".");
+    }
+
+    /**
+     * Tests that extracting a jar using {@code -P} flag and without any explicit destination
+     * directory works correctly if the jar contains entries with leading slashes and/or {@code ..}
+     * parts preserved.
+     */
+    @Test
+    public void testExtractNoDestDirWithPFlag() throws Exception {
+        // create a jar which has leading slash (/) and dot-dot (..) preserved in entry names
+        final Path jarPath = createJarWithPFlagSemantics();
+        // extract with -P flag without any explicit destination directory (expect the extraction to work fine)
+        final String[] args = new String[]{"-xvfP", jarPath.toString()};
```

Comment:
Perhaps add a DataProvider so you can test --extract as well?

Reply:
Not a data provider but the latest version of PR tests --extract as well